pkg/reconciler/common/prune.go

```diff
@@ -27,535 +27,592 @@ import (
-	//in this case delete all the cronjobs with the label pruneCronLabel
-	if pruner.removedFromTektonConfig(tC.Spec.Pruner) {
-		cronJobs, err := pruner.listCronJobs(ctx, tC.Spec.TargetNamespace, pruneCronLabel)
-		if err != nil {
-			return err
-		}
-		if len(cronJobs.Items) > 0 {
-			for _, cronJob := range cronJobs.Items {
-				if err := pruner.deleteCronJob(ctx, cronJob.Name); err != nil {
-					return err
-				}
-			}
-		}
-		return nil
-	}
-
-	//may be reconciled by removing/adding annotation on a Namespace
-	// if schedule is removed or prune.skip is added we delete pre-existing cron.
-	annotationRemovedUniqueCron, err := pruner.checkAnnotationsRemovedNamespaces(ctx)
+func Prune(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) error {
+	pruner, err := getPruner(ctx, k, tektonConfig)
 	if err != nil {
 		return err
 	}
-	if len(annotationRemovedUniqueCron) > 0 {
-		for _, uniqueCron := range annotationRemovedUniqueCron {
-			if err := pruner.checkAndDeleteCron(ctx, uniqueCron); err != nil {
-				return err
-			}
-		}
-	}
 
-	tknImageFromEnv := os.Getenv(JobsTKNImageName)
-	if tknImageFromEnv == "" {
-		return fmt.Errorf("%s environment variable not found", JobsTKNImageName)
-	}
-	pruner.tknImage = tknImageFromEnv
-	logger := logging.FromContext(ctx)
+	return pruner.reconcile()
+}
 
-	// for the default config from the tektonconfig
-	pruningNamespaces, err := prunableNamespaces(ctx, k, tC.Spec.Pruner)
-	if err != nil {
-		return err
-	}
-	if pruningNamespaces.commonScheduleNs != nil && len(pruningNamespaces.commonScheduleNs) > 0 {
-		jobs := getJobContainer(generateAllPruneConfig(pruningNamespaces.commonScheduleNs), pruner.targetNamespace, pruner.tknImage)
-		if err := pruner.checkAndCreate(ctx, "", tC.Spec.Pruner.Schedule, jobs, pruneCronLabel, pruningNamespaces.configChanged, tC); err != nil {
-			logger.Error("failed to create cronjob ", err)
-		}
-	}
+func getPruner(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) (*Pruner, error) {
+	pruner := &Pruner{
+		tektonConfig:    tektonConfig,
+		kubeClientset:   k,
+		targetNamespace: tektonConfig.Spec.TargetNamespace,
+		ownerRef:        *metav1.NewControllerRef(tektonConfig, tektonConfig.GetGroupVersionKind()),
+		ctx:             ctx,
+		logger:          logging.FromContext(ctx),
+	}
+	return pruner, nil
+}
 
-	if pruningNamespaces.uniqueScheduleNS != nil {
-		for ns, con := range pruningNamespaces.uniqueScheduleNS {
-			jobs := getJobContainer(generatePruneConfigPerNamespace(con, ns), ns, pruner.tknImage)
-			listOpt := pruneCronNsLabel + "=" + ns
-			if err := pruner.checkAndCreate(ctx, ns, con.config.Schedule, jobs, listOpt, pruningNamespaces.configChanged, tC); err != nil {
-				logger.Error("failed to create cronjob ", err)
-			}
-		}
+func (pr *Pruner) reconcile() error {
+	// get tkn cli container image name from environment
+	tknImageFromEnv := os.Getenv(prunerContainerImageEnvKey)
+	if tknImageFromEnv == "" {
```

Comment:
this is not required right

if there is an env, we set that image. If there is no env, we set the default whatever is there in the cronjob

Reply:
we can just add that as tranformer, no need to do validation for this

Reply:
@piyush-garg We need this. This is the only place we are getting the image detail and adding same into CronJob. there is no default image for pruner. It is mandatory to have environment `IMAGE_JOB_PRUNER_TKN` (prunerContainerImageEnvKey)